File: src/types/ai.ts
import type { Severity } from './common';

export type ChatRole = 'user' | 'assistant' | 'system';

export interface ChatCitation {
  id: string;
  label: string;
  type: 'event' | 'incident' | 'indicator' | 'asset' | 'log';
  href?: string;
}

export interface ChatMessage {
  id: string;
  role: ChatRole;
  content: string;
  timestamp: string;
  /** Terminal-flavoured output blocks rendered inside the message. */
  blocks?: { kind: 'timeline' | 'table' | 'metrics'; rows: string[][] | Record<string, string> }[];
  citations?: ChatCitation[];
  confidence?: number;
  status?: 'streaming' | 'complete' | 'error';
  error?: string;
}

export interface InvestigationContext {
  incidentId?: string;
  incidentTitle?: string;
  severity?: Severity;
  threatId?: string;
  source?: string;
  target?: string;
  attachedEventIds: string[];
  timeWindow: string;
}

export interface AiSession {
  id: string;
  title: string;
  createdAt: string;
  context: InvestigationContext;
  messages: ChatMessage[];
}

export interface SuggestedPrompt {
  id: string;
  label: string;
  prompt: string;
}

File: backend/app/schemas/ai.py
from typing import List, Optional, Union, Dict, Any
from pydantic import BaseModel
from datetime import datetime


class ChatCitation(BaseModel):
    id: str
    label: str
    type: str
    href: Optional[str] = None


class ChatBlock(BaseModel):
    kind: str
    rows: Union[List[List[str]], Dict[str, str]]


class ChatMessageResponse(BaseModel):
    id: str
    role: str
    content: str
    timestamp: str
    blocks: Optional[List[ChatBlock]] = None
    citations: Optional[List[ChatCitation]] = None
    confidence: Optional[float] = None
    status: Optional[str] = None
    error: Optional[str] = None


class InvestigationContext(BaseModel):
    incidentId: Optional[str] = None
    incidentTitle: Optional[str] = None
    severity: Optional[str] = None
    threatId: Optional[str] = None
    source: Optional[str] = None
    target: Optional[str] = None
    attachedEventIds: List[str] = []
    timeWindow: str = ""


class InvestigateRequest(BaseModel):
    question: str
    context: InvestigationContext